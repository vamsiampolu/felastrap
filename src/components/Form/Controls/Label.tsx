import { PropsWithChildren } from "react";
import { useFela } from "react-fela";

const labelRule = () => ({
  display: "inline-block",
  marginBottom: "0.5rem",
});

export type LabelProps = PropsWithChildren<{
  htmlFor: string;
  className?: string;
}>;

export const Label = ({ htmlFor, children, className = "" }: LabelProps) => {
  const { css } = useFela();
  return (
    <label
      data-testid="form-label"
      htmlFor={htmlFor}
      className={`${css(labelRule)} form-label ${className}`.trim()}
    >
      {children}
    </label>
  );
};
