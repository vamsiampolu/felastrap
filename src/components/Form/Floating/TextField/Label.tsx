import type { ComponentPropsWithoutRef } from "react";
import { useFela } from "react-fela";
import { useFloatingTextContext } from "./useFloatingTextContext";
import { StyleProps } from "../../../common";

export type RuleProps = StyleProps<{
  disabled: boolean;
  placeholder: boolean;
  focused: boolean;
  plainText: boolean;
  readOnly: boolean;
  autoFill: boolean;
}>;

const labelRule = () => ({
  display: "inline-block",
  marginBottom: "0.5rem",
  border: "1px solid transparent",
  height: "100%",
  left: 0,
  overflow: "hidden",
  padding: "1rem 0.75rem",
  pointerEvents: "none",
  position: "absolute",
  textAlign: "start",
  textOverflow: "ellipsis",
  top: 0,
  transformOrigin: "0 0",
  transition: "opacity 0.1s ease-in-out, transform 0.1s ease-in-out",
  whiteSpace: "nowrap",
  zIndex: 2,
  "@media (prefers-reduced-motion: reduce)": {
    transition: "none",
  },
});

const disabledRule = ({ disabled }: RuleProps) =>
  disabled
    ? {
        color: "#6c757d",
        "::after": {
          backgroundColor: "#e9ecef",
        },
      }
    : {};

const noPlaceholderRule = ({ placeholder }: RuleProps) =>
  placeholder
    ? {
        transform: "scale(0.85) translateY(0.5rem) translateX(0.15rem)",
      }
    : {
        color: "rgba(33, 37, 41, 0.65)",
        transform: "scale(0.85) translateY(-0.5rem) translateX(0.15rem)",
        "::after": {
          position: "absolute",
          inset: "1rem 0.375rem",
          zIndex: -1,
          height: "1.5em",
          content: "",
          backgroundColor: "#fff",
          borderRadius: "0.375rem",
        },
      };

const focusRule = ({ focused }: RuleProps) =>
  focused
    ? {
        color: "rgba(33, 37, 41, 0.65)",
        transform:
          "scale(0.85) translateY(-0.5rem) translateX(0.15rem) !important",
        "::after": {
          position: "absolute",
          inset: "1rem 0.375rem",
          zIndex: -1,
          height: "1.5em",
          content: "",
          backgroundColor: "#fff",
          borderRadius: "0.375rem",
        },
      }
    : {};

const plainTextReadOnlyRule = ({ plainText, readOnly }: RuleProps) =>
  plainText && readOnly
    ? {
        borderWidth: "1px 0",
        color: "rgba(33, 37, 41, 0.65)",
        transform: "scale(0.85) translateY(-0.5rem) translateX(0.15rem)",
        "::after": {
          position: "absolute",
          inset: "1rem 0.375rem",
          zIndex: -1,
          height: "1.5em",
          content: "",
          backgroundColor: "#fff",
          borderRadius: "0.375rem",
        },
      }
    : {};

const autofillRule = ({ autoFill }: RuleProps) => {
  if (autoFill) {
    return {
      color: "rgba(33, 37, 41, 0.65)",
      transform: "scale(0.85) translateY(-0.5rem) translateX(0.15rem)",
    };
  }

  return {};
};

export type FloatingLabelProps = Omit<
  ComponentPropsWithoutRef<"label">,
  "children"
> & { label?: string };

export const FloatingLabel = ({
  className = "",
  label,
  htmlFor,
  ...props
}: FloatingLabelProps) => {
  const {
    autoFilled: autoFill = false,
    disabled = false,
    focused = false,
    placeholderShown: placeholder = false,
    plainText = false,
    readOnly = false,
  } = useFloatingTextContext();

  const { css } = useFela({
    autoFill,
    disabled,
    focused,
    placeholder,
    plainText,
    readOnly,
  });

  const labelClassName = `${css([
    labelRule,
    focusRule,
    disabledRule,
    noPlaceholderRule,
    plainTextReadOnlyRule,
    autofillRule,
  ])} floating-label ${className}`.trim();

  return (
    <label
      data-testid="floating-label"
      htmlFor={htmlFor}
      className={labelClassName}
      {...props}
    >
      {label}
    </label>
  );
};
