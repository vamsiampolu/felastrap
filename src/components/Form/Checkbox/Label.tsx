import { useFela } from "react-fela";
import { LabelProps, Label } from "../Controls/Label";
import { StyleProps } from "../../common";

export type CheckboxLabelProps = LabelProps & { disabled?: boolean };

const disabledRule = ({ disabled }: StyleProps<{ disabled: boolean }>) =>
  disabled
    ? {
        cursor: "default",
        opacity: 0.5,
      }
    : {};

export const CheckboxLabel = ({
  disabled = false,
  className = "",
  htmlFor,
  children,
}: CheckboxLabelProps) => {
  const { css } = useFela({ disabled });
  const cx = `${css(disabledRule)} form-check-label ${className}`.trim();
  return (
    <Label htmlFor={htmlFor} className={cx}>
      {children}
    </Label>
  );
};
