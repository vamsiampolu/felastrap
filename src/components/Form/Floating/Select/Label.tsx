import { useFela } from "react-fela";
import { Label, LabelProps } from "../../Controls/Label";
import { StyleProps } from "../../../common";

const rule = () => ({
  position: "absolute",
  top: 0,
  left: 0,
  zIndex: 2,
  height: "100%",
  padding: "1rem 0.75rem",
  overflow: "hidden",
  textAlign: "start",
  textOverflow: "ellipsis",
  whiteSpace: "nowrap",
  pointerEvents: "none",
  border: "1px solid transparent",
  transformOrigin: "0 0",
  transition: "opacity 0.1s ease-in-out, transform 0.1s ease-in-out",
  color: "rgba(33, 37, 41, 0.65)",
  transform: "scale(0.85) translateY(-0.5rem) translateX(0.15rem)",
  "@media (prefers-reduced-motion: reduce)": {
    transition: "none",
  },
  "::after": {
    position: "absolute",
    inset: "1rem 0.375rem",
    zIndex: -1,
    height: "1.5em",
    content: "",
    backgroundColor: "#fff",
    borderRadius: "0.375rem",
  },
});

const disabledRule = ({ disabled }: StyleProps<{ disabled: boolean }>) =>
  disabled
    ? {
        color: "#6c757d",
      }
    : {};

export type FloatingSelectLabelProps = LabelProps & { disabled?: boolean };

export const FloatingSelectLabel = ({
  disabled = false,
  children,
  className = "",
  ...props
}: FloatingSelectLabelProps) => {
  const { css } = useFela({ disabled });
  const cx = `${css([
    rule,
    disabledRule,
  ])} floating-select-label ${className}`.trim();
  return (
    <Label className={cx} {...props}>
      {children}
    </Label>
  );
};
