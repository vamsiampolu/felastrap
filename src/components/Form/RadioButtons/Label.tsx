import { useFela } from "react-fela";
import { Label, type LabelProps } from "../Controls/Label";
import { StyleProps } from "../../common";

const disabledRule = ({ disabled }: StyleProps<{ disabled: boolean }>) =>
  disabled ? { cursor: "default", opacity: "0.5" } : {};

export type RadioLabelProps = LabelProps & { disabled?: boolean };

export const RadioLabel = ({
  className = "",
  children,
  htmlFor,
  disabled = false,
}: RadioLabelProps) => {
  const { css } = useFela({ disabled });
  const cx = `${css(disabledRule)} form-check-label ${className}`.trim();
  return (
    <Label className={cx} htmlFor={htmlFor}>
      {children}
    </Label>
  );
};
