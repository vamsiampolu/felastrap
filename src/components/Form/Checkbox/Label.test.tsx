import { IRenderer } from "fela";
import { CheckboxLabel } from "./Label";
import {
  Wrapper,
  createRenderer,
  createSnapshot,
  cssReboot,
} from "../../../test-helpers";
import { render, screen } from "@testing-library/react";

describe("components/Form/Checkbox/Label", () => {
  let renderer: IRenderer;

  beforeEach(() => {
    renderer = createRenderer();
    cssReboot(renderer);
  });

  afterEach(() => {
    renderer.clear();
  });

  it("should render a label", async () => {
    const component = (
      <CheckboxLabel htmlFor="cb-label">Checkbox Label</CheckboxLabel>
    );
    render(<Wrapper renderer={renderer}>{component}</Wrapper>);

    expect(screen.getByTestId("form-label")).toBeInTheDocument();
    expect(screen.queryByTestId("form-label")).toHaveAttribute(
      "for",
      "cb-label",
    );

    const snapshot = await createSnapshot({ component, renderer });
    expect(snapshot).toMatchSnapshot();
  });
  it("should render a label for a disabled checkbox", async () => {
    const component = (
      <CheckboxLabel htmlFor="cb-label" disabled>
        Checkbox Label
      </CheckboxLabel>
    );

    render(<Wrapper renderer={renderer}>{component}</Wrapper>);
    expect(screen.getByTestId("form-label")).toBeInTheDocument();

    const snapshot = await createSnapshot({ component, renderer });
    expect(snapshot).toMatchSnapshot();
  });
});
