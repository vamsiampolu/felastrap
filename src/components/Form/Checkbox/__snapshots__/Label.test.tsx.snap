// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Form/Checkbox/Label should render a label 1`] = `
".a {
  display: inline-block;
}
.b {
  margin-bottom: 0.5rem;
}


<label
  data-testid="form-label"
  for="cb-label"
  class="a b form-label form-check-label"
  >Checkbox Label</label
>;
"
`;

exports[`components/Form/Checkbox/Label should render a label for a disabled checkbox 1`] = `
".a {
  cursor: default;
}
.b {
  opacity: 0.5;
}
.c {
  display: inline-block;
}
.d {
  margin-bottom: 0.5rem;
}


<label
  data-testid="form-label"
  for="cb-label"
  class="c d form-label a b form-check-label"
  >Checkbox Label</label
>;
"
`;
